File: typescript/packages/core/src/evals/drift.ts
import type { DryPin } from "../agent/registry";
import type { EventOf } from "../fold/state";
import { canonicalize, parseStrictJson } from "../log";
import type { DriftCheck, ToolDiff } from "./schema";

// Drift (spec lane 22, B.3): the case's recorded config against the agent as it is pinned now,
// by a dry pin that runs no setup. The recorded prompt is line 0 of the case's first request;
// the model is the case's first settings epoch. What the dry pin can't see is `unchecked`, never
// stale.

/** A tool as pinned: the recorded one or the dry pin's draft of it. */
type Spec = {
  readonly name: string;
  readonly origin?: { readonly extension: string } | undefined;
};
/** A thread_started's data as drift reads it, recorded or freshly pinned. */
type Started = Pick<
  EventOf<"thread_started">["data"],
  "agent_name" | "config_hash" | "instructions"
> & {
  readonly parent?: { readonly relation: keyof typeof RELATIONS } | undefined;
  readonly model: unknown;
  readonly model_params: unknown;
  readonly adapter: unknown;
  readonly tools: readonly Spec[];
};

export type Recorded = {
  /** The case log's thread_started: the first settings epoch and the pinned tools. */
  readonly started: Started;
  /** line0.json (or a recorded request's line 0), when the case has one. */
  readonly line0: Uint8Array | undefined;
};

const RELATIONS = {
  subagent: "spawn",
  team_member: "team_member",
  handoff: "handoff",
} as const;

const text = (value: unknown): string => {
  const c = canonicalize(JSON.parse(JSON.stringify(value ?? null)));
  if (!c.ok) throw new Error("pinned values are canonical JSON");
  return c.value;
};

/** A pinned tool as drift compares it: everything but where it came from. */
const shape = ({ origin: _, ...spec }: Spec): string => text(spec);

/** The recorded line 0's system text, or undefined when there is no readable line 0. */
function recordedSystem(line0: Uint8Array | undefined): string | undefined {
  if (line0 === undefined) return undefined;
  const parsed = parseStrictJson(new TextDecoder().decode(line0));
  if (!parsed.ok || typeof parsed.value !== "object" || parsed.value === null)
    return undefined;
  const system: unknown = Reflect.get(parsed.value, "system");
  return typeof system === "string" ? system : undefined;
}

/** What the dry pin couldn't compare, as the report names it. */
function uncheckedOf(pin: DryPin): readonly string[] {
  return [
    ...pin.mcp.map((s) => `mcp:${s}`),
    ...pin.setupExtensions.map((e) => `extension:${e}`),
    ...pin.setupProviders,
  ];
}

/** Tools the dry pin can compare: no MCP tool, none a setup-bearing extension contributed. */
function comparable(
  tools: readonly Spec[],
  pin: DryPin,
): ReadonlyMap<string, string> {
  const setup = new Set(pin.setupExtensions);
  return new Map(
    tools
      .filter(
        (t) =>
          !pin.mcp.some((s) => t.name.startsWith(`mcp__${s}__`)) &&
          !setup.has(t.origin?.extension ?? ""),
      )
      .map((t) => [t.name, shape(t)]),
  );
}

function toolDiff(recorded: Started, pin: DryPin): ToolDiff | undefined {
  const before = comparable(recorded.tools, pin);
  const after = comparable(pin.started.tools, pin);
  const added = [...after.keys()].filter((n) => !before.has(n)).toSorted();
  const removed = [...before.keys()].filter((n) => !after.has(n)).toSorted();
  const changed = [...after]
    .filter(([n, s]) => before.has(n) && before.get(n) !== s)
    .map(([n]) => n)
    .toSorted();
  return added.length + removed.length + changed.length === 0
    ? undefined
    : { added, removed, changed };
}

/** A case saved before tool origins can't tell a setup-bearing extension's tools apart. */
const originless = (recorded: Started, pin: DryPin): boolean =>
  pin.setupExtensions.length > 0 &&
  !recorded.tools.some((t) => t.origin !== undefined);

function compare(recorded: Recorded, pin: DryPin): DriftCheck {
  const unchecked = [...uncheckedOf(pin)];
  const system = recordedSystem(recorded.line0);
  if (system === undefined) unchecked.push("no_recorded_prefix");
  const kinds: DriftCheck["kinds"] = [];
  const promptBlind =
    pin.setupExtensions.length > 0 || pin.setupProviders.length > 0;
  if (
    !promptBlind &&
    system !== undefined &&
    system !== pin.started.instructions
  )
    kinds.push("prompt");
  const tools = originless(recorded.started, pin)
    ? undefined
    : toolDiff(recorded.started, pin);
  if (tools !== undefined) kinds.push("tools");
  const settings = (s: Started) => text([s.model, s.model_params, s.adapter]);
  if (settings(recorded.started) !== settings(pin.started)) kinds.push("model");
  if (
    kinds.length === 0 &&
    unchecked.length === 0 &&
    recorded.started.config_hash !== pin.started.config_hash
  )
    kinds.push("config");
  return {
    ok: kinds.length === 0,
    kinds,
    ...(tools === undefined ? {} : { tools }),
    ...(unchecked.length === 0 ? {} : { unchecked }),
  };
}

/** The drift check of one case against the agents given. */
export function drift(recorded: Recorded, pins: readonly DryPin[]): DriftCheck {
  const relation = recorded.started.parent?.relation;
  if (relation !== undefined)
    return {
      ok: true,
      kinds: [],
      unchecked: [`relation:${RELATIONS[relation]}`],
    };
  const pin = pins.find(
    (p) => p.started.agent_name === recorded.started.agent_name,
  );
  if (pin === undefined)
    return {
      ok: false,
      kinds: [],
      agents: pins.map((p) => p.started.agent_name),
    };
  return compare(recorded, pin);
}

File: python/src/threads/evals/drift.py
"""Drift (spec lane 22, B.3): the case's recorded config against the agent as it is pinned now, by
a dry pin that runs no setup. The recorded prompt is line 0 of the case's first request; the model
is the case's first settings epoch. What the dry pin can't see is `unchecked`, never stale."""

import json
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Final

from pydantic import JsonValue

from threads.agents.definition import DryPin
from threads.evals.compare import canonical

type Started = Mapping[str, JsonValue]
"""A thread_started's data as wire JSON, recorded or freshly pinned."""

_RELATIONS: Final[Mapping[str, str]] = {
    "subagent": "spawn",
    "team_member": "team_member",
    "handoff": "handoff",
}


@dataclass(frozen=True, slots=True)
class Recorded:
    started: Started
    """The case log's thread_started: the first settings epoch and the pinned tools."""
    line0: bytes | None
    """line0.json (or a recorded request's line 0), when the case has one."""


@dataclass(frozen=True, slots=True)
class DriftResult:
    """The report's drift check, as fields in its key order."""

    ok: bool
    kinds: tuple[str, ...]
    tools: Mapping[str, tuple[str, ...]] | None = None
    unchecked: tuple[str, ...] | None = None
    agents: tuple[str, ...] | None = None

    def to_json(self) -> JsonValue:
        out: dict[str, JsonValue] = {"ok": self.ok, "kinds": list(self.kinds)}
        if self.tools is not None:
            out["tools"] = {k: list(v) for k, v in self.tools.items()}
        if self.unchecked is not None:
            out["unchecked"] = list(self.unchecked)
        if self.agents is not None:
            out["agents"] = list(self.agents)
        return out


def _tools(started: Started) -> list[Mapping[str, JsonValue]]:
    tools = started.get("tools")
    return [t for t in tools if isinstance(t, dict)] if isinstance(tools, list) else []


def _shape(spec: Mapping[str, JsonValue]) -> str:
    """A pinned tool as drift compares it: everything but where it came from."""
    return canonical({k: v for k, v in spec.items() if k != "origin"})


def _origin(spec: Mapping[str, JsonValue]) -> str:
    origin = spec.get("origin")
    extension = origin.get("extension") if isinstance(origin, dict) else None
    return extension if isinstance(extension, str) else ""


def _recorded_system(line0: bytes | None) -> str | None:
    """The recorded line 0's system text, or None when there is no readable line 0."""
    if line0 is None:
        return None
    try:
        parsed: JsonValue = json.loads(line0)
    except ValueError:
        return None
    system = parsed.get("system") if isinstance(parsed, dict) else None
    return system if isinstance(system, str) else None


def _unchecked(pin: DryPin) -> list[str]:
    return [
        *(f"mcp:{s}" for s in pin.mcp),
        *(f"extension:{e}" for e in pin.setup_extensions),
        *pin.setup_providers,
    ]


def _comparable(started: Started, pin: DryPin) -> dict[str, str]:
    """Tools the dry pin can compare: no MCP tool, none a setup-bearing extension contributed."""
    out: dict[str, str] = {}
    for t in _tools(started):
        name = str(t.get("name"))
        mcp = any(name.startswith(f"mcp__{s}__") for s in pin.mcp)
        if not mcp and _origin(t) not in pin.setup_extensions:
            out[name] = _shape(t)
    return out


def _tool_diff(recorded: Started, pin: DryPin) -> Mapping[str, tuple[str, ...]] | None:
    before, after = _comparable(recorded, pin), _comparable(pin.started, pin)
    added = tuple(sorted(n for n in after if n not in before))
    removed = tuple(sorted(n for n in before if n not in after))
    changed = tuple(sorted(n for n, s in after.items() if n in before and before[n] != s))
    if not (added or removed or changed):
        return None
    return {"added": added, "removed": removed, "changed": changed}


def _originless(recorded: Started, pin: DryPin) -> bool:
    """A case saved before tool origins can't tell a setup-bearing extension's tools apart."""
    return bool(pin.setup_extensions) and not any("origin" in t for t in _tools(recorded))


def _settings(s: Started) -> str:
    return canonical([s.get("model"), s.get("model_params"), s.get("adapter")])


def _compare(recorded: Recorded, pin: DryPin) -> DriftResult:
    unchecked = _unchecked(pin)
    system = _recorded_system(recorded.line0)
    if system is None:
        unchecked.append("no_recorded_prefix")
    kinds: list[str] = []
    blind = bool(pin.setup_extensions or pin.setup_providers)
    if not blind and system is not None and system != pin.started.get("instructions"):
        kinds.append("prompt")
    tools = None if _originless(recorded.started, pin) else _tool_diff(recorded.started, pin)
    if tools is not None:
        kinds.append("tools")
    if _settings(recorded.started) != _settings(pin.started):
        kinds.append("model")
    same_hash = recorded.started.get("config_hash") == pin.started.get("config_hash")
    if not kinds and not unchecked and not same_hash:
        kinds.append("config")
    return DriftResult(not kinds, tuple(kinds), tools, tuple(unchecked) if unchecked else None)


def drift(recorded: Recorded, pins: Sequence[DryPin]) -> DriftResult:
    """The drift check of one case against the agents given."""
    parent = recorded.started.get("parent")
    relation = parent.get("relation") if isinstance(parent, dict) else None
    if isinstance(relation, str):
        return DriftResult(True, (), unchecked=(f"relation:{_RELATIONS[relation]}",))
    name = recorded.started.get("agent_name")
    pin = next((p for p in pins if p.started.get("agent_name") == name), None)
    if pin is None:
        return DriftResult(False, (), agents=tuple(str(p.started.get("agent_name")) for p in pins))
    return _compare(recorded, pin)
